Aliasing == and != with explicit names

thomasblanc opened the pull request:
I like to spend time on [Stackoverflow's OCaml corner](https://stackoverflow.com/questions/tagged/ocaml). One mistake I repetitively see from beginners is them using the `==` operator instead of `=`. (see [[1]](https://stackoverflow.com/questions/58619586/expression-type-float-list-list-list-but-was-expected-of-type-float-list-list-b) [[2]](https://stackoverflow.com/questions/58141500/how-to-find-an-element-in-a-list-at-a-random-position) [[3]](https://stackoverflow.com/questions/57915875/ocaml-type-simple-recursive-function) [[4]](https://stackoverflow.com/questions/56318299/fizbuzz-in-ocaml-gets-gets-error-is-not-compatible-with-type-unit) [[5]](https://stackoverflow.com/questions/55773485/why-doesnt-this-code-work-recursive-function-gets-an-error) [[6]](https://stackoverflow.com/questions/55131535/printing-ocaml-function-after-compiling) and that's just the last few ones)

Of course, those newcomers usually only try to compare integers and therefore don't realize their mistake, and I pretty much always end up commenting something like "not the answer to your question but you shouldn't use `==`, or you will have a very bad surprise some day".

Unfortunately, the rise of Reason and the constant increase of newcomers only makes this problem more likely to cause serious bugs to happen in production OCaml code. That's why I'm proposing to deprecate the (much loved) `==` and `!=` operators.

Problems:
 * This PR adds a new name to `Stdlib`. I know this is bad. I don't think that the name I chose would collide with existing names, but that's possible. Of all the modules in the stdlib, perhaps only `Gc` could make sense hosting those new names, I don't think it is worth creating a new module though.
 * I count about 300 occurrences of `==` in trunk right now. I will change those if I get agreement in principle and a commonly accepted name.

thomasblanc committed:
Deprecating == and != in favor of explicit names

gasche commented:
I like the idea. I think that `Sys` would be a better place than `Gc`; it makes sense given that some implementations of OCaml give a different semantics to physical equality.

gasche commented:
For a callsite I think that `physically_equal v1 v2` reads better than `physical_equality v1 v2`, or maybe `phys_equal`.

yallop commented:
I like the idea, too, although I think a warning would do the job as well as deprecation: it would alert beginners to the pitfall without actually breaking old code.

thomasblanc commented:
> I like the idea. I think that `Sys` would be a better place than `Gc`; it makes sense given that some implementations of OCaml give a different semantics to physical equality.

It isn't really a system call. But since `opaque_identity` is there I guess it makes sense.

> For a callsite I think that `physically_equal v1 v2` reads better than `physical_equality v1 v2`, or maybe `phys_equal`.

`physically_equal` looks good to me. I think it's worth avoiding abbreviations like `phys` for such a function.

> I like the idea, too, although I think a warning would do the job as well as deprecation: it would alert beginners to the pitfall without actually breaking old code.

A deprecation warning doesn't break old code (unless they're compiled with the wrong set of options). This allows for new code that do need physical equality to use it without seeing a warning.

dbuenzli commented:
> It isn't really a system call.

It's increasingly understood that `Sys` stands for runtime system rather than operating system.

mshinwell commented:
I vote for `phys_equal`: its name is already established in `Base`.

yallop commented:
> A deprecation warning doesn't break old code 

That's true, but deprecation is often the first step on the path to removal.  Is the warning not meant to carry that implication here?

thomasblanc commented:
> > A deprecation warning doesn't break old code
> 
> That's true, but deprecation is often the first step on the path to removal. Is the warning not meant to carry that implication here?

In the case of Pervasives, I think it might stay here for a while. I mean, I have no idea of when `&` and `or` where marked as deprecated and they're still in there.

yminsky commented:
This sounds great. We did this years ago and it was a significant
improvement.

+1 for Mark's suggestion of phys_equal, which is the name in Base.

y

On Wed, Oct 30, 2019, 12:15 PM Thomas Blanc <notifications@github.com>
wrote:

> A deprecation warning doesn't break old code
>
> That's true, but deprecation is often the first step on the path to
> removal. Is the warning not meant to carry that implication here?
>
> In the case of Pervasives, I think it might stay here for a while. I mean,
> I have no idea of when & and or where marked as deprecated and they're
> still in there.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ocaml/ocaml/pull/9080?email_source=notifications&email_token=AAFOUJR3HC2YKQ2OXTIOKRLQRGXJ5A5CNFSM4JG27EAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOECU2AVA#issuecomment-547987540>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOUJVMIYZWK42532SP4VLQRGXJ5ANCNFSM4JG27EAA>
> .
>

gadmm commented:
Here is the result of running `grep -r --include \*.ml --include \*.mli -e '[a-zA-Z( ][!=]=[a-zA-Z) ]'` on opam sources: https://gitlab.com/gadmm/stdlib-experiment/raw/master/other/async_audit/phys_equal (2,16 MB). It confirms that the change can help catch improper uses, but the number of occurrences raises the question of managing the transition.

As mentioned above, deprecation warnings are relied on by many people to know when to take urgent action to avoid code breakage (e.g. all dune users with default settings). If there is no intention to break code then a warning that people can disable selectively can help, but an automated transition tool might be better.

gadmm commented:
Here's an idea. Instead of `phys_equal`, one could choose the names `(===)` and `(!==)`. Indeed:
* Replacing `==` with `===` and `!=` with `!==` in a codebase is much simpler: it can be done reliably enough with a regular expression and then checked by hand.
* The issue seems to be less their status as infix operators, than the collision with ubiquitous syntax from other languages (including ReasonML as you mention) that lets us think that `==` and `!=` should be the go-to comparison operators (and, in my experience, is confusing when context-switching between languages frequently).
* The widespread use of `==` and `!=` shows that perhaps it is worth keeping the convenience of an Stdlib infix operator.
* This new syntax will also come naturally for people used to == and !=. (As I discovered afterwards it also happens to be the syntax chosen by ReasonML.)

One drawback is that the name `===` (but not `!==`) seems to be a recurring idiom for a custom comparison operator (often structural, and rarely in public interfaces): https://gitlab.com/gadmm/stdlib-experiment/raw/master/other/async_audit/triple_equal. The only library I found exporting `===` is https://github.com/bennn/assertions. This does not affect backwards-compatibility, but I wonder if replacing `==` with `===` makes `==` available for that purpose or if people will have to turn to `====` (!) or something else.

As for the separate warning, let me expand: it could be a different alert than the one used for deprecation, and intended to explicitly mean that no code is going to break, so that there is a chance to find consensus in the community to make this warning non-fatal. (A variant on this idea is to have versioning information for the deprecated alert as suggested in another thread, and to make non-fatal when the deadline is explicitly set as “never”, but the discussions seemed to hint at more work required to implement versioning.)

dbuenzli commented:
This problem is likely best solved by *not* adding new operators. However if an operator has to be added I propose `=@=&=@=` which looks nice and is a tribute to various address-of operators.

gadmm commented:
@dbuenzli,

The operator vs. explicit name question is debatable. However, I find your contribution very low-effort: you state an opinion together with a disparaging comment.

You do this once again, after another low-effort reply at https://github.com/ocaml/ocaml/issues/9071#issuecomment-547220160, where you disparaged dune's attempt at making deprecation warnings meaningful as the "opinions" of a "third-party", when it was a choice by members of the OCaml community and they took care to provide an argumented rationale. You then misrepresented the complex situation of #9071 to better fit your opinion, by making statements that are in contradiction with findings of people who actually took care to look at previous discussions of the feature and at uses of the code in the wild. You also made it sound like I did not propose an alternative solution to the blocking part of the issue.

The form does not contribute to the discussions. Please no longer reply to my posts in this way. You have shown to me many times that you know how to put your programming experience to good use to the benefit of the community, and I usually enjoy your contributions.

On the contents, you make it sound like it should be obvious that it is more important not to add new operators than trying to provide a reasonable evolution path. This does not sound reasonable to me.

Anybody should mind that the 26000 occurrences for what is searchable publicly in opam amounts to 45% more than `Pervasives` (18000 as of Aug 2019). The deprecation of Pervasives was seen by many as a mistake, even though it was supposed to be fixable by a simple search-and-replace. It led many to wonder about the meaning of backwards compatibility and deprecation in OCaml. I would have preferred if the proponents of deprecation were also the ones to propose an assessment of the impact and a way to manage the transition.

dbuenzli commented:
> The operator vs. explicit name question is debatable. However, I find your contribution very low-effort: you state an opinion together with a disparaging comment.

Sorry I didn't think I needed to add a wall of text to explain that programming language have been struggling with notions of equalities for quite some time now and have used various inconsistent notations to denote them.  So if you want to avoid confusion either add an operator no one uses (like the one I proposed which I would still like, it's cute) or just use named identifiers (my preference). And yes design is often *opinions* whether you like them or not.

I will disregard your other comments since they are not fit for that discussion. But I will just mention that I did not misrepresent anything in the discussion you mentioned and `dune` is not making deprecations in anyway "meaningful": they are simply equating them with `-warn-error` in development. In other words they make deprecations hard errors. Good for them if they want it that way but I don't think their choice should prevent upstream from deprecating things.

That being said I don't think deprecating now `==` is fit, as usual first introduce the new path wait a bit and then start deprecating and eventually remove or not the deprecated feature. I have proposed more than once a protocol for deprecating things. It seems the dev team still keeps on shying away from the problem which ends up in endless debate about the process each time someone wants to deprecate something.

thomasblanc commented:
As for a switch to `===`, I'd say I'm not a huge fan. It does fix the problem at hand, but it will come back the moment someone coming from a language that has a `===` operator starts using it without realising it. Php is plagued with equality operators and if there is something we don't want, it's to be like php.

To me, the goals are:

1. Make sure unknowing users use this feature without realizing it.
1. Not break things.
1. Fix the problem for good and not have to come back to it.

To me, having an operator instead of a fully qualified name is a threat to both points 1 and 3. If you're one typo away for the same error to happen, then we're not really protected from errors. See [here](https://xkcd.com/1306/) for more details.

As for the ideas shared on deprecation, I'd say that I'm not really attached to it, but I think there should be *some kind* of message displayed whenever an unknowing user types `==` and that message should shipped into OCaml as soon as possible.

As for what the policy for deprecation should be, I don't feel like my opinion would be relevant. If one is given I would obviously follow it (and I guess dune would adapt to it).

dbuenzli commented:
> As for a switch to `===`, I'd say I'm not a huge fan.

It just makes it as bad as what there was before. It reintroduces the exact problem you were trying to get rid of, namely newcomer confusion on equality operators, except on other cases. Fire up a browser console and type  `1.0 === 1.0` and `"bla" === "bla"`.

I will repeat my low-effort but usability wise comment again, the problem is unlikely to be solved by adding new operators.

gadmm commented:
Thank you for stating design goals and constraints. I proposed the following goal:

4. Give an easy migration path.

Even with an informational message, it needs to be actionable by developers who want to get rid of it. What is proposed in the case of an explicit name? In other threads, it has been suggested that a tool similar to Rust's `cargo fix` could help in that sort of situation, but I am not aware of concrete plans to materialize it.

It is suggested that the new operators could confuse Javascript and PHP programmers. I don't really think it is an important threat, but if people think so, remember that they are also accustomed to dealing with different kinds of comparison operators. For 3. I agree that this is a middle ground that needs to be accepted so that we do not come back to it later.

dbuenzli commented:
@thomasblanc given the current state of affairs as far as deprecation is concerned your choices are either to:

1. Try to use regular identifier deprecation. In that case you won't be able to get warnings immediately. You should first introduce the named identifier without declaring the operators 
as deprecated (because otherwise people who need to compile code with old and new versions of the compiler get the warning on old versions without a good way of getting rid of it: if they silence it they will be caught by the eventual removal, if it happens, or will never migrate to the new identifier). Once the version that introduced the named operators is old enough you can introduce the operator deprecation (with the risk since there's no established procedure, that people will disagree on deprecating it at that point again).
2. You devise, as @yallop suggests, a new warning that warns on `==` usages and entices to use 
the named identifiers (though people that need to compile with multiple versions will also have no choice but to silence the warning on their code base). This being understood that adding a new warning and make it the default is also fraught with politics.

Number 2. likely fits more your goals. But in any case I'd suggest to you to separate the introduction of the named identifier for physical equality from the business of warnings and deprecations. It seems quite a few people are interested in having it and there's quite a consensus on where and with which name that should be. Having that in asap can only help you move towards your goal.

gasche commented:
I also prefer to have an explicit name, `phys_equal` or `Sys.phys_equal`, over an operator. I think that physical equality is a subtle operation, used by people who care about the performance implication of their code, so it deserves to have a longer, more explicit name, to better express intent and call attention (to both the reader and the writer) to this advanced feature.

If people write certain programs where physical equality is heavily used (I myself use it extremely rarely in typical code, but I suppose that certain hash-consing approaches may lead to its somewhat frequent use), they are free to give a name locally that they find more convenient, it could be `(===)` as suggested by Guillaume or even `(==)`.

I think that adding a new, good, explicit name for physical equality could be the focus of a first PR, as @dbuenzli suggests. No matter what deprecation or warning mechanism we decide to use, it's probably going to get displayed by users only gradually, and we need to have the "proper choice" available in advance for a smooth transition.

xavierleroy commented:
> physical equality is a subtle operation, used by people who care about the performance implication of their code,

Physical equality is essential for the correctness of many algorithms over graphs and other mutable data structures, e.g. in-place unification.  It's not an obscure thing that only performance freaks should use and that should be hidden in the `Sys` module: physical equality is the correct equality for mutable data types.  Please keep this in mind in this discussion.

mshinwell commented:
I believe `Base` has `phys_equal` exposed in the toplevel namespace.  This doesn't seem unreasonable to me given what @xavierleroy says.

xavierleroy commented:
I'm still hesitating between `phys_equal` and an infix operator, probably `===`.  I think we should gather more opinions (beyond those that have already been vehemently exposed here), in particular from authors of libraries of imperative data structures, who will have the most porting work to do.

thomasblanc commented:
> Physical equality is essential for the correctness of many algorithms over graphs and other mutable data structures, e.g. in-place unification. It's not an obscure thing that only performance freaks should use and that should be hidden in the `Sys` module: physical equality is the correct equality for mutable data types. Please keep this in mind in this discussion.

@xavierleroy Agreed, that's why I think it's worth adding a value in `Stdlib`. Given that `Base` uses `phys_equal` in the toplevel namespace (see [here](https://ocaml.janestreet.com/ocaml-core/latest/doc/base/Base/index.html#val-phys_equal)) and given that `===` is more likely to be custom-defined, I'd be more inclined towards `phys_equal` (also, I like words).

Do you have an opinion on the (vehemently debated) kind of message that should be displayed to the user?

jberdine commented:
> I think we should gather more opinions

FWIW, I know some code bases (involving mutable data structures) where changing `==` to `===` would be acceptable, but the other options would necessitate some hack to "undeprecate" `==`.

Also, `!=` is just as important, so I think we'd need something like `!==` or `=!=` depending on what mnemonic people have in mind for `!=`.

thomasblanc commented:
> But the other options would necessitate some hack to "undeprecate" `==`.

I was thinking maybe we should add a mechanism to indicate when a specific identifier was deprecated and an option to only trigger the deprecation warning on identifiers older than a specific version. That would be some work for some more fine-grained control over warnings and alerts. @jberdine is `-alert -deprecated` not working for you?

gadmm commented:
Here is a list of packages sorted by occurrence count if you wonder who to ask. I cut at 40 occurrences to keep it short but I imagine the threshold for annoyance is lower. (I removed an obvious false positive, but keep in mind that there are probably other false positives.)
```
1236 frama-c.19.0
1171 merlin.3.3.1
1147 frama-c-base.15.0
1125 coq.8.9.1
1125 coqide.8.9.1
859 why.2.41
615 flowtype.0.94.0
597 proverif.2.00
595 statverif.1.97pl1.1
557 flow_parser.0.80.0
464 bsbnative.1.9.4
439 opa-base.1.1.0+4263
408 haxe.3.4.7
359 cil.1.7.3
345 goblint-cil.1.7.4
313 alt-ergo.2.3.0
313 altgr-ergo.2.3.0
313 alt-ergo-lib.2.3.0
313 alt-ergo-parsers.2.3.0
286 doc-ock.1.2.1
279 odoc.1.4.1
249 elpi.1.4.0
238 alt-ergo-free.2.0.0
231 libres3.1.3
229 mldonkey.3.1.6
221 cduce.0.5.5
216 memcad.1.0.0
216 pgsolver.4.1
215 phox.0.89.170929
198 profiler-plugin.1.30
198 satML-plugin.1.30
198 fm-simplex-plugin.1.30
173 ocamlformat.0.9.1
151 goblint.1.0.0
147 why3-ide.1.2.0
147 why3-coq.1.2.0
147 why3.1.2.0
144 libra-tk.1.1.2
143 KaSim.4.0.0
139 why3-base.0.88.3
138 herdtools7.7.52
122 sundialsml.3.1.1p0
115 zipperposition.1.5
104 spdiff.0.1
101 camlp5.7.06.10-g84ce6cc4
94 rtop.3.4.0
94 reason.3.4.0
90 glsurf.3.3.1
89 coccinelle.1.0.2
86 relit-reason.0.0.2
83 ocamlnet.4.1.6
82 genet.0.6
81 beluga.0.8.2
80 orsetto.1.0
77 batteries.2.9.0
74 menhir.20190620
71 cubicle.1.1.2
67 diy.5.01
67 stog.0.18.0
64 eliom.6.7.0
63 grenier.0.7
58 chamo.2.03
57 links-postgresql.0.8
57 links.0.8
56 reason-parser.2.0.0
56 zenon.0.8.4
55 lwt_glib.1.1.0
55 lwt_camlp4.1.0.0
54 lwt_ppx.1.2.2
54 lwt_react.1.1.2
54 lwt.4.2.1
53 obrowser.1.1.1
53 javalib.3.0
52 argot.1.1
50 logtk.0.8.1
48 owl-zoo.0.5.0
48 owl-top.0.5.0
48 owl-base.0.5.0
48 owl-plplot.0.5.0
48 owl.0.5.0
47 js_of_ocaml-compiler.3.4.0
47 js_of_ocaml-lwt.3.4.0
47 earley.2.0.0
47 js_of_ocaml.3.4.0
47 js_of_ocaml-ppx_deriving_json.3.4.0
47 js_of_ocaml-toplevel.3.4.0
47 js_of_ocaml-ppx.3.4.0
47 cryptodbm.0.84.2
47 js_of_ocaml-tyxml.3.4.0
45 TCSLib.0.3
45 ocp-pp.1.99.19-beta
45 ocp-build.1.99.20-beta
45 archsat.1.0
44 js_of_ocaml-camlp4.3.1.0
44 js_of_ocaml-ocamlbuild.3.1.0
42 h2-lwt.0.3.0
42 h2-lwt-unix.0.3.0
42 omake.0.10.3
42 h2-mirage.0.3.0
42 h2.0.3.0
41 hpack.0.2.0
41 ppx_cstubs.0.1.0
41 sawja.1.5.6
```

alainfrisch commented:
I don't think that the choice should be based on the difficulty of migrating to the new syntax.  Given how pervasives physical equality is, we (the community) will want to provide a migration tool anyway, and the extra effort to have this tool support a named value as opposed to an operator is rather small (the only tricky part is to decide whether to put parentheses around arguments, but a conservative heuristics -- parentheses always except around simple identifiers and constants would be good enough). We (LexiFi) have gone through such syntactic migration internally (loading .cmt files and rewriting .ml files in place), and @nojb wrote a migration tool; if needed, I think we could contribute the tool for rewriting == to phys_eq or whatever is decided.

gadmm commented:
@alainfrisch, sounds exciting. Having such a tool could unlock the discussion.

trefis commented:
> `1171 merlin.3.3.1`

FTR: my own unsophisticated `ag "==" -G ".*\.mli?"  | wc -l` only returns 1087 at the tip of our git repository. However, merlin contains a verbatim copy of the compiler frontend for each version since 4.02 in an `upstream` directory at the root of the repository, as well a patched copy of each of these version under `src/ocaml`.
If you take that into account, I'm happy to report that we're below your threshold by a large margin:
```
$ ag "==" -G ".*\.mli?" --ignore upstream --ignore src/ocaml | wc -l
20
```
:tada: 


And now that I have an excuse to join the discussion: I am really not excited about having a `===`.

Gbury commented:
Also for the record : the regexp used for the list seems to also capture `==` signs that are in strings (for instance in debug printing/logging), potentially yielding quite a few false positives.

gadmm commented:
(Yes, as I mentioned, there are false positives. I checked at a glance that it did not affect the orders of magnitudes, but I really do not have the time to refine the data further.)

jberdine commented:
> is `-alert -deprecated` not working for you

Good point. It would depend on whether an alert id is used that is specific enough to disable without other undesired consequences.

thomasblanc committed:
removing deprecation

thomasblanc commented:
After a discussion with the powers that be, I've removed the deprecation indicators. I also renamed the functions to `phys_equal` and `phys_inequal`.

I let them in `Stdlib`. We can keep the flame war active on that subject.

thomasblanc committed:
renaming to phys_equal: fix documentation

yminsky commented:
Very minor, but I think the English reads better as phys_unequal, or alternately, not_phys_equal (or maybe, phys_not_equal?) inequal is archaic, unequal is preferred, I believe.

jberdine commented:
FWIW, it is not uncommon in logic to use 'disequality' to refer to the negation of an 'equality', to distinguish it from an 'inequality' that asserts a less-than ordering.

gasche commented:
And we could also consider having only the equality and using `not`.

thomasblanc commented:
@yminsky Pardon my French :-p unless there is a strong opinion against it I'll change to unequal (the "not" doesn't convince me)

@gasche provided my intent is to eventually replace `==` and `!=`, I'd disagree with that.

gasche commented:
I'm confused, what's wrong with suggesting to replace `foo != bar` with `not (phys_equal foo bar)`?

thomasblanc commented:
@gasche I don't like making my parse-tree deeper. I find the complexity of writing code to usually rise with the number of brackets opened. That is, of course, a matter of personal taste. `!=` might as well be considered not-usefull-enough to get its own name.

alainfrisch commented:
Not that I particular like this style, but I've seen people writing `if not @@ foo bar then`, precisely to avoid parentheses.

yminsky commented:
Unless there's a compelling performance reason to keep it, I would prefer not having phys_unequal, or whatever you call it. We have String.equal and not String.unequal. I think we should follow the same pattern here.  For what it's worth, we've deprecated == and != for a long time in our codebase, and only have phys_equal, and its never felt like a problem.

thomasblanc committed:
Removing phys_inequal

thomasblanc committed:
removing phys_inequal from pervasives.ml

thomasblanc commented:
It is removed.

A summary of what may still be controversial IMHO:
 * `phys_equal` has received the most support and little opposition so that's the name I'm going for
 * I put it in `Stdlib`, since as @xavierleroy pointed out this is basic mutable value manipulation
 * The documentation comment is moved from `==` to `phys_equal`, with `==` being noted as an alias. One may want the opposite

gasche commented:
We could have a documentation comment for `( == )` that explains the reason for having an alias:

```
Language beginners often use this operator by mistake, influenced by the syntax
common in other languages. To clarify your intent, we recommend using
the more explicit `phys_equal` function instead.
```

yawaramin commented:
Hello, I skimmed through this discussion and saw a couple of mentions of 'alert'. Just to clarify: can we not use [alerts](https://ocaml.org/manual/alerts.html) here? Has this already been discussed? It seems to me an obvious solution–deprecation has the wrong semantics but alerting has the correct one–we just want to alert users they may be doing something wrong. I.e.

```ocaml
external ( == ) : 'a -> 'a -> bool = "%eq"
[@@alert phys_equal "Language beginners often use this operator by mistake, influenced by the syntax common in other languages. To clarify your intent, we recommend using the more explicit `phys_equal` function instead."]
```

And of course, this specific alert can be turned on or off in a fine-grained way.

thomasblanc commented:
Oh, I have to complete this PR at some point.

IMHO an alert is a bad option if there is no way to distinguish between a good usage and a bad one. Experienced devs should be able to code with physical equality without having to turn off an alert. A check that only activates the alert if `'a` shouldn't be tested for physical equality would be ideal, but explicit names that can't be confused by beginners in the first place seems like a better solution as it voids the problem entirely.

yawaramin commented:
It solves half of the problem as it leaves `==` and `!=` as they are now, usable without any warnings. People will ignore `phys_equal` and use what looks familiar, `==`. Experienced devs can still code with physical equality without having to turn off an alert by using `phys_equal` (which is what gasche's wording recommends), and if they don't want to do that, they can pay a small price for the convenience of using infix `==` by turning off the alert in their local scope.

Octachron reviewed (approved):
I think that introducing `phys_equal` as the new canonical name for `==` is a good first step.

 @thomasblanc : there is some bookkeeping left (writing a change entry, rebasing the PR) that I am proposing to take care of, except if you object.

gasche commented:
I was reminded of this issue by a new article about OCaml making the rounds today, [7 OCaml gotchas](https://dev.to/chshersh/7-ocaml-gotchas-207e). Can we move forward and add `Stdlib.phys_equal` as an alias to `==`?

(Let's wait a couple days to find out, but my guess is that @thomasblanc will not be available to work on this. Any volunteer to adopt the PR and bring it to partial completion?)

yawaramin commented:
By the way, alerts are disabled by default. If we added the `phys_equal` alert, it wouldn't actually do anything unless the project explicitly turned it on. So it would just be a nice easy way to audit possible equality mistakes.

Of course, a human would still have to check and fix each alert one by one. But it would be easier than manually scanning through the code looking for the `( == )` and `( != )` operators.

Octachron commented:
I am closing this PR since the operators `==` and `!=` have now explicit names in the `Repr` module.
